src/dmd/dsymbolsem.d

```diff
@@ -3775,6 +3884,23 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
                         ctd.deprecation("all parameters have default arguments, "~
                                     "but structs cannot have default constructors.");
                 }
+                else if ((dim == 1 || (dim > 1 && tf.parameterList[1].defaultArg)) &&
+                        ctd.ident != Id.copyCtor)
+                {
+                    //printf("tf: %s\n", tf.toChars());
+                    auto param = Parameter.getNth(tf.parameterList, 0);
+                    if (param.storageClass & STC.ref_ && param.type.mutableOf().unSharedOf() == sd.type.mutableOf().unSharedOf())
+                    {
+                        //printf("copy constructor\n");
```

Comment:
I don't get what is going on here. Somehow, this is going from detecting a copy constructor to generating one? Why is it generating one here?

Reply:
Whenever a constructor is identified as a copy constructor, a copy constructor definition is created. This is done for 2 reasons:

1. The constructor functions need to have a different name then the copy constructors. This makes it easier to resolve overloads and also avoids such situations:

```d
struct A
{
     int* p;
     this(A rhs) {}                          // 1
     this(ref A rhs) immutable {}   // 2
}

void main()
{
     A a;
     A b = a;  // rewritten to  b.__ctor(a) => will resolve to 1 bypassing the cpCtor
}
```

2. extern(C++) on a copy constructor affects the mangling of the constructor, but is ignored for the copy constructor.

Reply:
1. The easiest way to deal with this is to make `this(A rhs)` illegal. (It's illegal in C++.)
2. This should be straightforward to handle with the `isCopyCtor` field.

The neat thing is copy constructors won't have to be overload resolved separately from regular constructors.

Reply:
>The easiest way to deal with this is to make this(A rhs) illegal. (It's illegal in C++.)

In this case it would have to be all qualified versions:  `this($q1 A rhs) $q2`. This may be a little restrictive, although I can't think of a case where defining such a constructor would be useful. Should this code error even if a copy constructor is not defined?

Reply:
> In this case it would have to be all qualified versions: this($q1 A rhs) $q2.

Yes, good point.

> This may be a little restrictive, although I can't think of a case where defining such a constructor would be useful.

Neither can I. Such code is most likely a newbie mistake.

> Should this code error even if a copy constructor is not defined?

Yes.

Reply:
This breaks phobos [1] and I suspect it will break more code out there.

[1] https://github.com/dlang/phobos/blob/master/std/typecons.d#L8957

Reply:
That constructor shouldn't be there. The only reason that I can think of for it's existence would be a workaround for a compiler bug. But regardless, we should try remove it.

Reply:
Having such constructors results in infinite recursion. The parameter is copied to the stack, which requires the copy constructor, which requires copying the parameter to the stack, which requires the copy constructor, etc.

I understand the concern about breaking code. One possibility is to refuse to allow copy construction in code that has such constructors, similarly to refusal when a postblit is present.

Reply:
Meanwhile, a PR to fix typecons should be submitted.

Reply:
>Having such constructors results in infinite recursion. The parameter is copied to the stack, which requires the copy constructor, which requires copying the parameter to the stack, which requires the copy constructor, etc.

That is not correct. Let's say we have:

```d
struct A
{
     this(A rhs) {}
     this(ref A rhs) {}
}
```

The first functions is a constructor, while the second is also a constructor and a copy constructor. You can imagine that there are 3 functions defined:

```d
struct A
{
     void __ctor(A rhs) {}                    // 1
     void __ctor(ref A rhs) {}               // 2
     void __copyCtor(ref A rhs) {}       // 3
}

ctor 1 will be called explicitly only on rvalues: A(fun()), where fun returns an object of type A. Since fun is in rvalue, the return value will be moved.
ctor 2 will be called excliptly only on lvalues: A(b) where b is a an object of type A. Note that the parameter is passed by reference, so no copy is employed.
ctor 3 will be called implicitly only on lvalues when a copy is performed (documented in the DIP).

So there is no possibility of infinite recursion.

Also, consider the following situation which has nothing to do with the current patch:

```d
import std.stdio : writeln;
struct A
{
    this(A a) { writeln("ctor"); }
}

void main()
{
     A a;
     A b = a;
}

For `A b = a` the compiler does not call the constructor, which means that when an initialization is done by copy the compiler either blits the data (if no postblit is present) or it calls the postblit (if present). Now consider this case:

```d
struct A
{
     int* p;
     this(immutable A rhs) {}
} 

immutable(A) fun() { immutable A a; return a; }

void main()
{
     A b = fun();
}
```

If A'a constructor becomes illegal, then there is no way to initialize an object from an rvalue with mutable indirections.

Reply:
When we fix `ref` to allow it to bind rvalues, the copy constructor will be to do its job properly ;)

Reply:
I think that the current model of using lvalues with ref is fine; we should let the user bind the rvalue to an lvalue and then call the function with it.

Reply:
>That's right, one would need to write a copy constructor to do it. That's the way it should be.
You cannot write a copy constructor for an rvalue. By the current rules, copy constructors work only on lvalues, so there is no copy constructor that you can define to make that work; you either have to explicitly bind the rvalue to an lvalue and then call the copy constructor on it, or you can do I mentioned and simply explicitly call the constructor for rvalues.

>if there is both a this(A) and this(ref A) then that is an error.

But things are not that simple. You may have this(immutable A) and this(ref A). These two constructors are meant for different things: the first is meant for constructing from an rvalue (explicitly), the second is meant for copy constructing (implicitly); it is only a matter of convenience that you can call the copy constructor explcitly as you would do with a normal constructor.  Banning this(immutable A) because you define this(ref A) just doesn't make sense  with the current language rules.

>This is just much more complicated than necessary. Nobody is going to understand the difference between 2 and 3

It really isn't that complicated. The difference between the 2 is that one is called explicilty while the other is implicitly. That's all. You can look at it this way: copy constructors can only be called on rvalues; for everything else you can just define a constructor as you would have done before the DIP. The only matter of confusion here is that we chose to define a copy constructor with the same syntax as a constructor, otherwise everything would have been more clearer.

>All the special cases, the need for naming them differently, the copying of the AST's, special cases with overload resolution, it all falls away.

There are no special cases. You rename them differently to make sure that constructors work exaclty the same as before this addition and you copy the AST to have them as different functions and that's all of it. In addition, the user does not know how it was implemented; for him, when he defines a copy constructor he defines both a constructor (which he can call explicitly) and a copy constructor (which is always called implicitly). I don'y see any reason of confusion here. If we want to make things clear, maybe we should just change the syntax and have copy constructors be a completely different thing from constructors.

Reply:
I missed this discussion originally. The reason one one want to have:

```d
struct Foo {
    this(ref Foo);
    this(Foo);
}
```

is simple: C++ interop. I made it so that the dpp translations actually enabled D code to call a C++ move constructor by overloading on this.